### PR TITLE
Review for USB Camera Device Service

### PR DESCRIPTION
**Pull request by iain-anderson**
This service is proposed for adoption into the main edgexfoundry project.
The process for doing so is documented here: https://wiki.edgexfoundry.org/display/FA/Process+for+new+Device+Services

### CODE REVIEW

File: README.md
Comment (mpvats-intel) [inline README.md]: Link should be (with the .example appended) --  https://github.com/edgexfoundry-holding/device-usb-camera/blob/main/cmd/res/devices/hp.w200.toml.example .. but, maybe minus the "-holding" as well since we'll be moving off of holding.

**Review by mpvats-intel — APPROVED** [review]

File: Dockerfile
Comment (lenny-goodell) [inline Dockerfile]: Change to port assigned by @iain-anderson

File: Makefile
Comment (lenny-goodell) [inline Makefile]: Please update to match new approach here: https://github.com/edgexfoundry/device-rfid-llrp-go/blob/main/Makefile#L23-L35

File: Dockerfile
Comment (lenny-goodell) [inline Dockerfile]: Move down after FROM alpine:3.15 to so it is in the built image not in the builder section

File: cmd/res/configuration.toml
Comment (lenny-goodell) [inline cmd/res/configuration.toml]: Remove. It isn't an example.

File: cmd/res/configuration.toml
Comment (lenny-goodell) [inline cmd/res/configuration.toml]:
```suggestion
# Must also add `ADD_SECRETSTORE_TOKENS: "device-usb-camera"`  and  `ADD_KNOWN_SECRETS: redisdb[device-usb-camera]` to secret-store-setup environment so it generates  the token and secret store in vault for 'device-usb-camera' and added the Redis credentials
```

File: cmd/res/configuration.toml
Comment (lenny-goodell) [inline cmd/res/configuration.toml]: Use consistent indentation. ie. remove one level here.

File: cmd/res/configuration.toml
Comment (lenny-goodell) [inline cmd/res/configuration.toml]: Remove indentation

File: docker-compose/README.md
Comment (lenny-goodell) [inline docker-compose/README.md:1]: Add TODO to remove these `docker-compose` files once service accepted into EdgeX Foundry and added to compose builder.

File: go.mod
Comment (lenny-goodell) [inline go.mod:9]: @iain-anderson , these new modules will need to be TSC approved.

File: internal/driver/constants.go
Comment (lenny-goodell) [inline internal/driver/constants.go]: Use more readable names for the constant names. No reason they have to be the same as the cryptic values

**Review by lenny-goodell — CHANGES_REQUESTED** [review]
First pass review (non-code) completed. Next will be the code files.

File: Dockerfile
Comment (lenny-goodell) [inline Dockerfile]:
```suggestion
COPY go.mod vendor* ./
RUN [ ! -d "vendor" ] && go mod download all || echo "skipping..."

COPY . .
```

File: cmd/res/configuration.toml
Comment (lenny-goodell) [inline cmd/res/configuration.toml]: Change to port assigned by @iain-anderson

File: internal/driver/driver.go
Comment (lenny-goodell) [inline internal/driver/driver.go]: Use Info since it is on start up
```suggestion
	d.lc.Infof("RTSP server hostname: %s", rtspServerHostName)
```

File: internal/driver/driver.go
Comment (lenny-goodell) [inline internal/driver/driver.go]:
```suggestion
	d.lc.Infof("RTSP TCP port: %s", rtspPort)
```

File: internal/driver/driver.go
Comment (lenny-goodell) [inline internal/driver/driver.go]: `v4l2` has no meaning/context here. Use better name
```suggestion
	cameraDevice, err := usbdevice.Open(device.path)
```

File: internal/driver/driver.go
Comment (lenny-goodell) [inline internal/driver/driver.go]:
```suggestion
	usbdevice "github.com/vladimirvivien/go4vl/v4l2/device"
```

File: internal/driver/driver.go
Comment (lenny-goodell) [inline internal/driver/driver.go]: Find why to reduce this duplicated code that is in every case. helper run to create error?

File: internal/driver/driver.go
Comment (lenny-goodell) [inline internal/driver/driver.go]: change variable name so it has meaning so easer to understand the code.
```suggestion
			if _, ok := currentDevices[cn+sn]; ok {
```

File: internal/driver/driver.go
Comment (lenny-goodell) [inline internal/driver/driver.go]: Shouldn't this be done first thing?

File: internal/driver/driver.go
Comment (lenny-goodell) [inline internal/driver/driver.go]:
```suggestion
	cameraDevice, err := v4l2device.Open(fdPath)
```

File: internal/driver/driver.go
Comment (lenny-goodell) [inline internal/driver/driver.go]:
```suggestion
	cameraDevice, err := v4l2device.Open(path)
```

File: internal/driver/metadata.go
Comment (lenny-goodell) [inline internal/driver/metadata.go]:
```suggestion
	usbdevice "github.com/vladimirvivien/go4vl/v4l2/device"
```

**Review by lenny-goodell — CHANGES_REQUESTED** [review]
Completed review of the code.

**Comment by iain-anderson** [comment]
Use port 59983 by default (configuration, Dockerfile)

File: README.md
Comment (FelixTing) [reply README.md]: Modified. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/b2d801d418cc0860212bd79b15c013c06d4194d9

File: Dockerfile
Comment (FelixTing) [reply Dockerfile]: Changed to 59983. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/27ef6ca5f58755da044353e6a3ba7fea6c19b0e2

File: Makefile
Comment (FelixTing) [reply Makefile]: Updated. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/e89199cd6cd4b8f97e6f5b4d065b3641450654e0

File: Dockerfile
Comment (FelixTing) [reply Dockerfile]: Modified. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/847b52c537cb9bf5e5e55f10b4bbc1ffd82892c5
(Also removed another redundant label)

File: cmd/res/configuration.toml
Comment (FelixTing) [reply cmd/res/configuration.toml]: Removed. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/1ff5a040929b4f505be2a5cb5db9e6c9b4afe5a5

File: cmd/res/configuration.toml
Comment (FelixTing) [reply cmd/res/configuration.toml]: Modified as per the suggested change. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/1ff5a040929b4f505be2a5cb5db9e6c9b4afe5a5

File: cmd/res/configuration.toml
Comment (FelixTing) [reply cmd/res/configuration.toml]: Removed one level of indentation. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/1ff5a040929b4f505be2a5cb5db9e6c9b4afe5a5

File: cmd/res/configuration.toml
Comment (FelixTing) [reply cmd/res/configuration.toml]: Removed. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/1ff5a040929b4f505be2a5cb5db9e6c9b4afe5a5

File: docker-compose/README.md
Comment (FelixTing) [reply docker-compose/README.md:1]: Added the TODO note. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/b2d801d418cc0860212bd79b15c013c06d4194d9

File: cmd/res/configuration.toml
Comment (FelixTing) [reply cmd/res/configuration.toml]: Changed to 59983. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/27ef6ca5f58755da044353e6a3ba7fea6c19b0e2

File: internal/driver/constants.go
Comment (FelixTing) [reply internal/driver/constants.go]: Submitted a PR for this modification: https://github.com/edgexfoundry-holding/device-usb-camera/pull/14

File: internal/driver/driver.go
Comment (FelixTing) [reply internal/driver/driver.go]: Changed severity level to Info. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/f947e30ba36bca58cf452328e83108c19d23a6f3

File: internal/driver/driver.go
Comment (FelixTing) [reply internal/driver/driver.go]: Changed severity level to Info. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/f947e30ba36bca58cf452328e83108c19d23a6f3

File: internal/driver/driver.go
Comment (FelixTing) [reply internal/driver/driver.go]: Modified. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/b915490962f8488701e51706647b02e88983813a

File: internal/driver/driver.go
Comment (FelixTing) [reply internal/driver/driver.go]: Modified. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/b915490962f8488701e51706647b02e88983813a

File: internal/driver/driver.go
Comment (FelixTing) [reply internal/driver/driver.go]: Modified. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/b915490962f8488701e51706647b02e88983813a

File: internal/driver/driver.go
Comment (FelixTing) [reply internal/driver/driver.go]: You are right. Modified. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/8414d18ef7d553f530fd641a84d2cf3a0311ec64

File: internal/driver/driver.go
Comment (FelixTing) [reply internal/driver/driver.go]: Modified. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/b915490962f8488701e51706647b02e88983813a

File: internal/driver/driver.go
Comment (FelixTing) [reply internal/driver/driver.go]: Modified. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/b915490962f8488701e51706647b02e88983813a

File: internal/driver/metadata.go
Comment (FelixTing) [reply internal/driver/metadata.go]: Modified. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/b915490962f8488701e51706647b02e88983813a

File: internal/driver/driver.go
Comment (FelixTing) [reply internal/driver/driver.go]: Added helper to create error. https://github.com/edgexfoundry-holding/device-usb-camera/pull/11/commits/6bf2eb7deee61de122435ec2a6cb7ef6c2a26f19

**Review by lenny-goodell — APPROVED** [review]
Approved pending TSC approval of these new modules:

- github.com/vladimirvivien/go4vl
- github.com/xfrr/goffmpeg

File: README.md
Comment (farshidtz) [inline README.md:56]:
```suggestion
* The commands started with **VIDEO_** prefix are related to video stream.
```

File: Makefile
Comment (farshidtz) [inline Makefile:6]: Line 5 missing a 1 in GO11MODULE.

This variable can be removed in both lines, because Go 1.16 turns GO111MODULE=on by default. Go 1.17 ignores it alltogether: https://go.dev/blog/go116-module-changes

```suggestion
GO=CGO_ENABLED=0 go
GOCGO=CGO_ENABLED=1 go
```

File: README.md
Comment (farshidtz) [inline README.md:90]: This should be boolean `false`, not string `"false"`. Maybe they are equivalent.
```
type Device struct {
	autoStreaming               bool
        ...
}
```

File: cmd/res/devices/hp.w200.toml.example
Comment (farshidtz) [inline cmd/res/devices/hp.w200.toml.example:10]: Same comment on the readme. It should be boolean `false`, not string `"false"`.

File: README.md
Comment (farshidtz) [inline README.md:175]:
```suggestion
To manually trigger a Dynamic Discovery, use this [device service API](https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/device-sdk/2.1.0#/default/post_discovery).
```

File: README.md
Comment (farshidtz) [inline README.md:165]: Shouldn't this have a unit to clarify that this isn't a number but a Go duration string? Same for the environment variable few lines below.
```
export DEVICE_DISCOVERY_INTERVAL=0
```

File: README.md
Comment (farshidtz) [inline README.md:176]: `30s` isn't a valid [ISO8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations). This looks like a [Go duration string](https://pkg.go.dev/time#ParseDuration).

File: docker-compose/README.md
Comment (farshidtz) [reply docker-compose/README.md:1]: It would be better to create a Github issue for this that gets transfered over and worked on. These temp files could also be on a separate branch.

File: docker-entrypoint.sh
Comment (farshidtz) [inline docker-entrypoint.sh:7]: The docker container runs this script as the main process. The script in turn calls the rtsp server in background and device-usb-camera as sub-process. Two concerns:
- What happens if rtsp server fails?
- Does docker stop, stop both processes?

Running multiple processes in a container isn't recommended. But if you really have to do it, the main process should be orchestrating the runtime of both. See https://docs.docker.com/config/containers/multi-service_container/

File: go.mod
Comment (farshidtz) [reply go.mod:9]: Neither of `github.com/vladimirvivien/go4vl` or `github.com/xfrr/goffmpeg` have a stable release.

File: internal/driver/driver.go
Comment (farshidtz) [inline internal/driver/driver.go:256]: The wait group is used in both this and the startStreaming functions. So call to Wait() waits for both streams to stop (lines 522 and 529) and StopStreaming to return. I think adding a comment here is useful to clarify this.

File: internal/driver/driver.go
Comment (farshidtz) [inline internal/driver/driver.go:618]: This doesn't capture standard errors nor does it return the standard output on error. Consider using [`CombinedOutput()`](https://pkg.go.dev/os/exec#Cmd.CombinedOutput) and return the output to the caller on error.

**Review by farshidtz — CHANGES_REQUESTED** [review]
I did a code review and it looks good overall. 

I've commented on few minor issues. ~~There is a potentially important issue in the docker-entrypoint.sh which needs to be addressed.~~ Also the dependencies are pernding approval.

File: docker-entrypoint.sh
Comment (farshidtz) [reply docker-entrypoint.sh:7]: It looks like dump-init interpreted set in shebang takes care of that. I did not notice it at first.
https://github.com/Yelp/dumb-init#session-behavior

There is this comment inside the dockerfile, which doesn't explain dump-init is used for handling multiple processes:
```
# dumb-init needed for injected secure bootstrapping entrypoint script when run in secure mode.
RUN apk add --update --no-cache zeromq dumb-init ffmpeg udev
```

File: README.md
Comment (lenny-goodell) [reply README.md:90]: @farshidtz the protocols are map[string]string so it have to be `"false"`

File: README.md
Comment (lenny-goodell) [reply README.md:165]: Good catch. Should be "1h" same as device-onvif-camera

File: README.md
Comment (lenny-goodell) [reply README.md:176]: Interesting. Yes it is a Go duration string.

File: README.md
Comment (MonicaisHer) [inline README.md:110]: Do you mean `"InputImageSize":"640x480"`? 
`"InputFrameSize"` is unsupported.

**Comment by iain-anderson** [comment]
Service has been approved and moved out of -holding.
@MonicaisHer @farshidtz  can you please create new issues to address your remaining review comments

**Pull request closed without merging**
